Generalize processors

rombrom opened the pull request:
When attempting local fork indexing I de-duplicated some things, figuring any EVM chain would (roughly?) be set up the same way. Additionally updated the indexer tooling a bit.

- Removed processors `src/eth-sepolia` and `src/polygon-mumbai`.
- Added generic processor in `src/processor.ts`.
- Added `src/processors/*` for sepolia, sepolia fork and mumbai.
- Modified our log parsers to utilize `Fields` configuration instead of requiring instance specific `typeof processor`.
- Updated `NETWORK` enum to support `eth-sepolia-fork`
- Added `simple-import-sort` for module imports.

rombrom committed:
chore: install `sqd` CLI and setup package scripts

rombrom committed:
feat: set up generalized processor factory

rombrom committed:
chore: move processors to `src/processors`

rombrom committed:
chore: clean up `src/consts.ts` and consuming modules

rombrom committed:
chore: add import sorting

rombrom committed:
chore: mark eslintrc as root

rombrom committed:
chore: formatting

rombrom committed:
fixup! chore: move processors to `src/processors`

rombrom committed:
fixup! chore: move processors to `src/processors`

rombrom committed:
feat: sepolia fork indexing

rombrom committed:
chore: remove `@subsquid/cli` and `concurrently`

rombrom committed:
feat: add `endBlock` option

rombrom committed:
fix: lowercase addresses, because else our `logParser` won't pick up things

rombrom committed:
chore: move `.toLowerCase()` clarification

nazariyv commented on src/processors/eth-sepolia-local.ts:
```diff
@@ -0,0 +1,24 @@
+import { start } from "../processor";
+
+start({
+  dbOptions: { stateSchema: "eth_sepolia_processor" },
```
`eth_sepolia_fork_processor`

nazariyv commented on src/processors/eth-sepolia-local.ts:
```diff
@@ -0,0 +1,24 @@
+  // am not sure about this, because squid is supposed to handle automatically,
+  // but if error happens, all we would have to do is re-run / re-deploy the
+  // squid. So re-deploying once a year (for sepolia) is fine.
+  finalityConfirmation: 6,
```
this isn't required along with the comment. there can be no re-orgs in anvil

nazariyv commented on src/processors/eth-sepolia-local.ts:
```diff
@@ -0,0 +1,24 @@
+  rentalFactoryAddress: "0x2c2BBA22aA19Ba34bC5BA65e6c35Ce54DA36A33D",
+  rentalManagerAddress: "0xea0b609F81B3D7699a970e670ec471Daf687E5c2",
+  source: {
+    // archive: lookupArchive("sepolia", { type: "EVM" }),
```
**note to self: need to make this processor two-stage:**

(i) process from archive for the range: from contract deployed up to the forked block

(ii) immediately after the above processor start a new one: indexes from anvil for the ranges: from forked block and without to block

also note: consider whether forked block number should be used inclusively or exclusively in the above. If you use archive to index up to and including that block and anvil contains the same state, you aren't running into any issues because indexer would not re-index, since it saves into the db what was the last block that it saw. As such, this should not yield any issues:

```
// first processor: archive indexer; range:
startBlock: <when contract was deployed>,
endBlock: <forked block number>
```

```
// second processor: anvil indexer; range:
startBlock: <forked block number>
```

rombrom replied:
I took a micro stab at this yesterday but it's kind of sucky. Because the processor takes over the process I couldn't just `await start(...` and run the second `start()`. Unsure if subsquid has control flow like this exposed. Probably not. A potential solution here would be to `bash` it or see if we can update the `commands.json`, which appears to be invoking shell commands also.

nazariyv replied:
going to leave this unresolved (interested in the startBlock / endBlock discussion; in case it resurfaces in the future). mentioned this conversation in fork readme.

nazariyv commented on src/processors/eth-sepolia-local.ts:
```diff
@@ -0,0 +1,24 @@
+    // archive: lookupArchive("sepolia", { type: "EVM" }),
+    chain: process.env.RPC_ETH_SEPOLIA_HTTP_LOCAL ?? "http://127.0.0.1:8545",
+  },
+  startBlock: 3923754,
```
**note to self:**
magic numbers

nazariyv commented on src/consts.ts:
```diff
@@ -13,37 +13,9 @@ export const FIELDS = {
+export type Fields = typeof FIELDS;
+
 export enum NETWORK {
   ETH_SEPOLIA = "eth-sepolia",
```
need `ETH_SEPOLIA_FORK = "eth-sepolia-fork"` so that it saves it into the db appropriately

need to change the language from `*-local` to `*-fork` everywhere else too

rombrom committed:
chore: update `dbOptions.stateSchema` declaration

rombrom committed:
chore: set Sepolia fork `finalityConfirmation` to `1`

rombrom committed:
chore: rename `*local*` to `*fork*` w/r Sepolia fork indexer

rombrom committed:
chore: add magic number todo

rombrom committed:
chore: add `NETWORK.ETH_SEPOLIA_FORK`

rombrom committed:
fix: no hard-coded `NETWORK.ETH_SEPOLIA` on `new Block`s `network`

rombrom committed:
fix: rename file to `eth-sepolia-fork.ts` and update `.env.example`

rombrom committed:
fix: actually commit the update to `NETWORK` enum

rombrom commented on src/mapping/rentalManager.ts:
```diff
@@ -121,7 +124,7 @@ export function parseFunction(
         blockNumber: transaction.block.height,
         blockHash: transaction.block.hash,
-        address: consts.CONTRACT_ADDRESS[chain][consts.CONTRACT.RENTAL_MANAGER],
+        address: contractAddress,
```
Was wondering whether we could use `transaction.to` here. Would shave off a parameter.

nazariyv replied:
will leave this unresolved. haven't looked at this, we can incorporate in the future PRs

nazariyv commented on src/processors/polygon-mumbai.ts:
```diff
@@ -0,0 +1,25 @@
+    archive: lookupArchive("polygon-mumbai", { type: "EVM" }),
+    chain: process.env.RPC_POLYGON_TESTNET_HTTP,
+  },
+  startBlock: 39817388,
```
magic numbers

nazariyv commented on src/processors/eth-sepolia.ts:
```diff
@@ -0,0 +1,31 @@
+    archive: lookupArchive("sepolia", { type: "EVM" }),
+    chain: process.env.RPC_ETH_SEPOLIA_HTTP,
+  },
+  startBlock: 3923754,
```
magic numbers

nazariyv commented on src/processors/eth-sepolia-fork.ts:
```diff
@@ -0,0 +1,23 @@
+    chain: process.env.RPC_ETH_SEPOLIA_FORK_HTTP ?? "http://127.0.0.1:8545",
+  },
+  // TODO: explain these magic numbers
+  startBlock: 3923754,
```
magic numbers

rombrom committed:
feat: make default `processor:eth-sepolia-fork` a two-step process

Removed `src/processors/eth-sepolia-fork.ts` and migrated the containing
configuration into `src/processors/eth-sepolia-fork/config.ts`. The
containing directory now houses two processors:

1. `src/processors/eth-sepolia-fork/archive.ts` which does an initial
   pass through the historical chain data. This process completes with
   exit code `0` when it hits the `endBlock`.
2. `src/processors/eth-sepolia-fork/chain.ts` which is the processor
   which starts to index the actual Sepolia fork and will run
   indefinitely.

The `commands.json` file has been updated so that:

- `npx @subsquid/cli process:eth-sepolia-fork` runs the Sepolia fork
  archive processor as a dependency. After this process completes it
  starts the Sepolia chain processor.
- `npx @subsquid/cli process:eth-sepolia-fork:archive` runs the Sepolia
  fork archive processor standalone.
- `npx @subsquid/cli process:eth-sepolia-fork:chain` runs the Sepolia
  fork chain processor standalone.

rombrom committed:
chore: add magic number TODOs

nazariyv committed:
chore(docs): added comments and readme

nazariyv committed:
chore: magic number comment

nazariyv committed:
chore(commands): docker-compose -> docker compose
